### PR TITLE
chore(docs): v3 docs update

### PR DESCRIPTION
**Pull request by dbanksdesign**
*Issue #, if available:* #571 #159 #605

*Description of changes:* WIP! 

* Renamed "style properties" to "design tokens" (up for discussion, just wanted to see how it felt)
    * Adding `tokens` and `allTokens` to the core internal data structure to be consistent with this change. This is purely additional, `properties` and `allProperties` still exist and this should not break existing code. Moving forward examples should use `tokens` rather than `properties`.
* properties.md -> tokens.md
    * Adding more content in this doc page about: 
    * All potential attributes #605
    * Default attributes Style Dictionary adds
    * More info on what categories the CTI currently supports #571
    * Information on file formats for token files, including JS modules #159, and custom parsers for token files
* Updated broken links (linking to "master" branch instead of "main")
* Updated the examples page with all new examples
* Added a page on custom parsers
* Added information about transitive transforms
* Removed "docs/build_process.md" in favor of "docs/architecture.md" because the content was mostly the same. 
* Cleaned up configuration content 
    * Updated language around config file formats and how to define the configuration
    * Added information about configuration attributes and separated it out into sections: top-level config attributes including customization, platform attributes, and file attributes, for readability
* Cleaning up some spacing and headers in markdown files 

Discussion points:
* "pre-defined" v. "built-in" with respect to pre-defined transforms and formats
* "design tokens" v. "properties"??
* "input" and "output" files, or "property files" and "generated files"? (as opposed to "source" files which might be confusing because there are "source" and "include")

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/build_process.md [hunk 1]
```diff
@@ -4,13 +4,25 @@ Here is what the build system is doing under the hood.
...
+    1. Perform any actions defined in the actions attribute
 
 # How to Build
+
 You can build a style dictionary [using the cli](using_the_cli.md) or [using the npm module](using_the_npm_module.md).
+
+
+## Parsers
+
+Starting in version 3.0, you can define custom parsers to parse property files. This allws you to define your property files in *any* language you like as long as you can write a parser for it.
```
Comment (dbanksdesign) [inline docs/build_process.md:22]: This section needs to be filled out and linked to the properties.md page custom parser section

Comment (dbanksdesign) [reply docs/build_process.md]: build_process and architecture are 90% the same. Decided to remove build_process, open to renaming architecture build_process though if it makes more sense.

File: README.md [hunk 2]
```diff
@@ -206,13 +215,13 @@ $ style-dictionary build
 Take a look at the documentation for the example code.
 
 
-## Style Properties
+## Design tokens
 
-A style property is an attribute to describe something visually. It is atomic (it cannot be broken down further). Style properties have a name, a value, and optional attributes or metadata. The name of a property can be anything, but we have a proposed naming structure that works really well in the next section.
+A design token is an attribute to describe something visually. It is atomic (it cannot be broken down further). Style properties have a name, a value, and optional attributes or metadata. The name of a token can be anything, but we have a proposed naming structure that works really well in the next section.
```
Comment (chazzmoney) [inline README.md:220]: > proposed naming structure that works really well

proposed naming structure that we find works really well

File: docs/formats.md [hunk 3]
```diff
@@ -59,7 +59,7 @@ A special file configuration is `filter`, which will filter the tokens before th
 
 * An object which gets passed to [Lodash's filter method](https://lodash.com/docs/4.17.14#filter).
 * A string that references the name of a registered filter, using the [`registerFilter`](api.md#registerfilter) method
-* A function if you are defining your configuration in Javascript rather than JSON. The filter function takes a token as the property and should return a boolean if the token should be included (true) or excluded (false).
+* A function if you are defining your configuration in Javascript rather than JSON. The filter function takes a token and should return a boolean if the token should be included (true) or excluded (false).
```
Comment (chazzmoney) [inline docs/formats.md:62]: >  A function if you are defining your configuration in Javascript rather than JSON. The filter function takes a token as the property and should return a boolean if the token should be included (true) or excluded (false).

A function that takes a token and returns a boolean if the token should be included (true) or excluded (false). ***This is only available if you are defining your configuration in Javascript.

Or something. Just trying to make it easier to parse.

File: docs/parsers.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# Parsers
+
+Starting in version 3.0, you can define custom parsers to parse design token files. This allows you to define your design token files in *any* language you like as long as you can write a parser for it. 
+
+A custom parser matches design token files based on a file path regular expression. It will get the contents of a file as a string and should return an object of the data. 
+
+Custom parsers can be used to keep design token files in YAML, but they can also be used to add extra metadata onto all properties before transforms are run or anything else you can think of modifying on the design tokens before transforms are run.
```
Comment (chazzmoney) [inline docs/parsers.md:7]: I think you are trying to share example use cases here, but this is hard to parse and a bit confusing.

File: docs/transforms.md [hunk 5]
```diff
@@ -4,7 +4,7 @@ EDIT scripts/handlebars/templates/api.hbs OR JSDOC COMMENT INSTEAD!
...
+Transforms are functions that transform a [design token](tokens.md) - this enables each platform to consume the design tokens in different ways. A simple example is changing pixel values to point values for iOS and dp or sp for Android. Transforms are applied in a non-destructive way thus each platform can transform the properties. The order you use transforms matters because transforms are performed sequentially. Transforms are used in your [configuration](config.md), and can be either [pre-defined transforms](transforms.md?id=defining-custom-transforms) supplied by Style Dictionary or [custom transforms](transforms.md?id=defining-custom-transforms).
```
Comment (chazzmoney) [inline docs/transforms.md:7]: Transforms are functions that modify a token so that it can be understood by a specific platform. It can modify the name, value, or attributes of a token.

Comment (chazzmoney) [reply docs/transforms.md:7]: Transforms are isolated per platform; each platform begins with the same design token and makes the modifications it sees fit without affecting other platforms.

File: docs/transforms.md [hunk 6]
```diff
@@ -25,20 +25,68 @@ A transform consists of 4 parts: type, name, matcher, and transformer. Transform
...
 
-**Name:** A name transform transform the name of a property. You should really only be apply one name transformer because they will override each other if you use more than one.
+**Name:** A name transform transform the name of a design token. You should really only be apply one name transformer because they will override each other if you use more than one.
 
-**Value:** The value transform is the most important as this is the one that changes the representation of the value. Colors can be turned into hex values, rgb, hsl, hsv, etc. Value transforms have a matcher function that filter which properties that transform runs on. This allows us to only run a color transform on only the colors and not every property.
+**Value:** The value transform is the most important as this is the one that changes the representation of the value. Colors can be turned into hex values, rgb, hsl, hsv, etc. Value transforms have a matcher function that filter which properties that transform runs on. This allows us to only run a color transform on only the colors and not every design token.
```
Comment (chazzmoney) [inline docs/transforms.md:32]: > Value transforms have a matcher function that filter which properties that transform runs on.

Value transforms have a matcher function that filter which tokens that transform runs on.

File: README.md [hunk 7]
```diff
@@ -206,13 +215,13 @@ $ style-dictionary build
...
+While not exactly necessary, we feel this classification structure of style properties makes the most sense semantically. Style properties can be organized into a hierarchical tree structure with the top level, category, defining the primitive nature of the token. For example, we have the color category and every token underneath is always a color. As you proceed down the tree to type, item, sub-item, and state, you get more specific about what that color is. Is it a background color, a text color, or a border color? What kind of text color is it? You get the point. Now you can structure your token json files like simple objects:
```
Comment (chazzmoney) [inline README.md:224]: > While not exactly necessary, we feel this classification structure of style properties makes the most sense semantically. Style properties can be organized into a hierarchical tree structure with the top level, category, defining the primitive nature of the token.

While not exactly necessary, we feel this classification structure of design tokens makes the most sense semantically. Design tokens can be organized into a hierarchical tree structure with the top level, category, defining the primitive nature of the token.

File: README.md [hunk 8]
```diff
@@ -2,17 +2,16 @@
...
-<br/>
-[![Build Status](https://img.shields.io/travis/amzn/style-dictionary.svg?style=flat-square)](https://travis-ci.org/amzn/style-dictionary)
+[![PRs welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](https://github.com/amzn/style-dictionary/blob/main/CONTRIBUTING.md#submitting-pull-requests)
+[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/amzn/style-dictionary/Test?style=flat-square)](https://github.com/amzn/style-dictionary/actions/workflows/test.yml)
 [![downloads](https://img.shields.io/npm/dm/style-dictionary.svg?style=flat-square)](https://www.npmjs.com/package/style-dictionary)
 
 # Style Dictionary
 > *Style once, use everywhere.*
 
 A Style Dictionary uses design tokens to define styles once and use those styles on any platform or language. It provides a single place to create and edit your styles, and exports these properties to all the places you need - iOS, Android, CSS, JS, HTML, sketch files, style documentation, etc. It is available as a CLI through npm, but can also be used like any normal node module if you want to extend its functionality.
```
Comment (chazzmoney) [inline README.md:12]: properties -> tokens

File: README.md [hunk 9]
```diff
@@ -132,9 +133,11 @@ This tells the style dictionary build system how and what to build. The default
...
 ```
+
 | Attribute | Type | Description |
 | :--- | :--- | :--- |
-| source | Array | Paths to the property json files. Can have globs. |
+| source | Array | An array of file path [globs](https://github.com/isaacs/node-glob) to design token files. Style Dictionary will do a deep merge of all of the token files, allowing you to organize your files files however you want. |
+| include | Array | An array of file path [globs](https://github.com/isaacs/node-glob) to design token files that contain default styles. The Style Dictionary uses this as a base collection of properties. The properties found using the "source" attribute will overwrite properties found using include. |
 | platforms | Object | Sets of platform files to be built. |
 | platforms | Array | Paths to the property json files. Can have globs. |
 | platform.transformGroup | String (optional) | Apply a group of transforms to the properties, must either define this or `transforms`. |
```
Comment (chazzmoney) [inline README.md:143]: properties -> tokens

File: README.md [hunk 10]
```diff
@@ -147,6 +150,7 @@ This tells the style dictionary build system how and what to build. The default
 | platform.file.options.showFileHeader | Boolean | If the generated file should have a "Do not edit + Timestamp" header (where the format supports it). By default is "true". |
 
 ### Properties
```
Comment (chazzmoney) [inline README.md:152]: properties -> tokens

File: docs/transforms.md [hunk 11]
```diff
@@ -25,20 +25,68 @@ A transform consists of 4 parts: type, name, matcher, and transformer. Transform
...
+{
+  "color": {
+    "red": { "value": "#f00" },
+    "danger": { "value": "{color.red}", "darken": 0.75 },
+    "error": { "value": "{color.danger}", "darken": 0.5 }
+  }
+}
+```
+
+Using a custom transitive transform you could have `color.danger` darken `color.red` and `color.error` darken `color.danger`. The pre-defined transforms are *not transitive* to be backwards compatible and not break anything.
```
Comment (chazzmoney) [inline docs/transforms.md:80]: > The pre-defined transforms are *not transitive* to be backwards compatible and not break anything.

The pre-defined transforms are *not transitive* to be backwards compatible with Style Dictionary 2.0 - an upgrade should not cause breaking changes.

File: README.md [hunk 12]
```diff
@@ -206,13 +215,13 @@ $ style-dictionary build
 Take a look at the documentation for the example code.
 
 
-## Style Properties
+## Design tokens
```
Comment (chazzmoney) [inline README.md:218]: tokens -> Tokens

File: scripts/handlebars/templates/transforms.hbs [hunk 13]
```diff
@@ -21,20 +21,68 @@ A transform consists of 4 parts: type, name, matcher, and transformer. Transform
...
+{
+  "color": {
+    "red": { "value": "#f00" },
+    "danger": { "value": "{color.red}", "darken": 0.75 },
+    "error": { "value": "{color.danger}", "darken": 0.5 }
+  }
+}
+```
+
+Using a custom transitive transform you could have `color.danger` darken `color.red` and `color.error` darken `color.danger`. The pre-defined transforms are *not transitive* to be backwards compatible and not break anything.
```
Comment (chazzmoney) [inline scripts/handlebars/templates/transforms.hbs:76]: > backwards compatible and not break anything.

backwards compatible with Style Dictionary 2.0 so an upgrade should not cause breaking changes.

**Comment by dbanksdesign** [comment]
Great comments @chazzmoney 

Next level shit

File: docs/config.md [hunk 14]
```diff
@@ -36,60 +72,77 @@ module.exports = {
...
+| source | Array[String] | An array of file path [globs](https://github.com/isaacs/node-glob) to design token files. Style Dictionary will do a deep merge of all of the token files, allowing you to organize your files files however you want. |
+| properties | Object | The properties object is a way to include inline style properties as opposed to using the `source` and `include` arrays. 
+| platforms | Object[Platform] | An object containing [platform](#platform) config objects that describe how the Style Dictionary should build for that platform. You can add any arbitrary attributes on this object that will get passed to formats and actions (more on these in a bit). This is useful for things like build paths, name prefixes, variable names, etc.
 
-----
+### Platform
+
+A platform is a collection of configuration that tells Style Dictionary how to transform and format your style properties. You can have as many platforms as you need and you can name them anything, there are no restrictions.
```
[the remark this answers is not in the document]
Comment (nhoizey) [reply docs/config.md:153]: Every time I read something about a "platform" here, I change it to "build target" in my head to better understand what it means. I have "multiple build targets" for the "Web platform", with Sass, CSS, JSON…

File: README.md (hunk 10 above)
Comment (nhoizey) [reply README.md:152]: Yes, please! 🙏

File: docs/config.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (dbanksdesign) [reply docs/config.md:153]: Oh nice @nhoizey! I like "build target"
